x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts

```diff
@@ -118,279 +119,329 @@ export const createLifecycleExecutor =
       InstanceContext,
       ActionGroupIds
     >
-  ) =>
-  async (
-    options: RuleExecutorOptions<
-      Params,
-      WrappedLifecycleRuleState<State>,
-      InstanceState,
-      InstanceContext,
-      ActionGroupIds
-    >
-  ): Promise<{ state: WrappedLifecycleRuleState<State> }> => {
-    const {
-      services: { alertFactory, shouldWriteAlerts },
-      state: previousState,
-      flappingSettings,
-      maintenanceWindowIds,
-      rule,
-    } = options;
-
-    const ruleDataClientWriter = await ruleDataClient.getWriter();
-
-    const state = getOrElse(
-      (): WrappedLifecycleRuleState<State> => ({
-        wrapped: previousState as State,
-        trackedAlerts: {},
-        trackedAlertsRecovered: {},
-      })
-    )(wrappedStateRt<State>().decode(previousState));
-
-    const commonRuleFields = getCommonAlertFields(options);
-
-    const currentAlerts: Record<string, ExplicitAlertFields> = {};
-    const alertUuidMap: Map<string, string> = new Map();
-
-    const lifecycleAlertServices: LifecycleAlertServices<
-      InstanceState,
-      InstanceContext,
-      ActionGroupIds
-    > = {
-      alertWithLifecycle: ({ id, fields }) => {
-        currentAlerts[id] = fields;
-        const alert = alertFactory.create(id);
-        const uuid = alert.getUuid();
-        alertUuidMap.set(id, uuid);
-        return alert;
-      },
-      getAlertStartedDate: (alertId: string) => state.trackedAlerts[alertId]?.started ?? null,
-      getAlertUuid: (alertId: string) => {
-        const uuid = alertUuidMap.get(alertId);
-        if (uuid) {
-          return uuid;
-        }
-
-        const trackedAlert = state.trackedAlerts[alertId];
-        if (trackedAlert) {
-          return trackedAlert.alertUuid;
-        }
-
-        const trackedRecoveredAlert = state.trackedAlertsRecovered[alertId];
-        if (trackedRecoveredAlert) {
-          return trackedRecoveredAlert.alertUuid;
-        }
-
-        const alertInfo = `alert ${alertId} of rule ${rule.ruleTypeId}:${rule.id}`;
-        logger.warn(
-          `[Rule Registry] requesting uuid for ${alertInfo} which is not tracked, generating dynamically`
-        );
-        return v4();
-      },
-      getAlertByAlertUuid: async (alertUuid: string) => {
-        try {
-          return await fetchAlertByAlertUUID(ruleDataClient, alertUuid);
-        } catch (err) {
-          return null;
-        }
-      },
-    };
-
-    const wrappedExecutorResult = await wrappedExecutor({
-      ...options,
-      state: state.wrapped != null ? state.wrapped : ({} as State),
-      services: {
-        ...options.services,
-        ...lifecycleAlertServices,
-      },
-    });
+  ) => {
+    let untrackLifecycleAlerts = () => {};
```

Comment:
This diff is deceptively large due to a bunch of indentation, so let me explain what happened here.

Instead of directly returning an `async function`, the `createLifecycleExecutor` function will first assign it to a `const` called `executor`.

Within the `executor` function, right at the end, it will create a function called `untrackLifecycleAlerts` and assign it to the `executor` object, using a slightly more convoluted equivalent to `this.untrackLifecycleAlerts = () => ...` to avoid Typescript errors.

The reason we do this is because `createLifecycleExecutor` is a function that gets used over a number of different plugins, and sometimes takes multiple different `ruleDataClient` objects within the same plugin for different rule types. Also, each plugin implements different helper functions and factories that call `createLifecycleExecutor`.

We need to use the exact same `ruleDataClient` to apply the Untracked state to a disabled rule's alerts, but disabling a rule disables the scheduled task to run its executor. So we need to create a new function that we can run in a different task, but still preserve all of the context of the original rule's executor. Hence, we can use the fact that you can just assign additional properties to functions, and add this new functionality to every plugin without having to refactor anything outside the core dependencies.

THIS IS ONE OF THE JAVASCRIPTEST THINGS I HAVE CODED IN YEARS AND I AM SO DEEPLY PROUD OF THIS ABSOLUTE HOT MESS OF A HACK